src/librustc_mir/const_eval.rs

```diff
@@ -106,17 +106,43 @@ pub fn mplace_to_const<'tcx>(
     let alloc = ecx.memory.get(ptr.alloc_id)?;
     assert!(alloc.align.abi() >= mplace.align.abi());
     assert!(alloc.bytes.len() as u64 - ptr.offset.bytes() >= mplace.layout.size.bytes());
-    // FIXME: only clone the parts that interest us (starting at offset, going to offset + size)
-    let mut alloc = alloc.clone();
-    // we take `mplace.layout.align` instead of `mplace.align`
-    // as this function is essentially copying the value
-    // out of the larger allocation, so we lose all information about
-    // potential surrounding types with different alignment.
-    alloc.align = mplace.layout.align;
+    // FIXME: stop cloning and refer to parts of allocations by giving `ConstValue::ByRef` fields
+    // for alignment overrides and size of the referred to part
+    let mut new_alloc = Allocation {
+        bytes: alloc
+            .bytes[ptr.offset.bytes() as usize..][..mplace.layout.size.bytes() as usize]
+            .to_owned(),
+        mutability: Mutability::Immutable,
+        relocations: Relocations::from_presorted(alloc
+            .relocations
+            .iter()
+            .filter_map(|&(offset, (tag, id))| if offset < ptr.offset {
+                None
+            } else {
+                Some((
+                    offset - ptr.offset,
+                    (tag, id)
+                ))
+            })
+            .collect()),
+        undef_mask: UndefMask::new(mplace.layout.size),
+        // we take `mplace.layout.align` instead of `mplace.align`
+        // as this function is essentially copying the value
+        // out of the larger allocation, so we lose all information about
+        // potential surrounding types with different alignment.
+        align: mplace.layout.align,
+    };
```

Comment:
All this memory-modifying code outside of `memory.rs` makes me really sad. :(  The (subtle and correctness-critical!) relocations invariant was already scattered across that entire file, can we avoid scattering it over the entire rustc codebase?